test/distributed/_composable/fsdp/test_fully_shard_init.py

```diff
@@ -41,5 +44,29 @@ def test_normalize_device_cuda(self):
                     self.assertEqual(_normalize_device(device), torch.device("cuda", 1))
 
 
+class TestFullyShardMeshArg(FSDPTestMultiThread):
+    """Tests the ``mesh`` argument."""
+
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    @unittest.skipIf(not TEST_CUDA, "no cuda")
+    def test_invalid_mesh_ndim(self):
+        mesh = init_device_mesh("cuda", (self.world_size, 1, 1))
+        model = MLP(8)
+        regex = r"fully\_shard expects a 1D or 2D DeviceMesh but got DeviceMesh\(\[\[\[0\]\], \[\[1\]\]\]\)"
+        with self.assertRaisesRegex(ValueError, regex):
+            fully_shard(model, mesh=mesh, device="cuda")
+
+    @unittest.skipIf(not TEST_CUDA, "no cuda")
+    def test_mesh_device_mismatch(self):
+        mesh = init_device_mesh("cuda", (self.world_size,))
+        model = MLP(8)
+        regex = "device and mesh must be of the same type but got cpu for device and cuda for mesh"
+        with self.assertRaisesRegex(ValueError, regex):
+            fully_shard(model, mesh=mesh, device="cpu")
```

Comment:
nit: we should add some tests that test that a correct mesh passes? or are those test would be added once the FSDP runtime logic is there?

Reply:
This is me being sloppy.

However, upon some more thought, I prefer to not add this test here. The reason is that I am not sure how we want to expose the mesh publicly from a `fully_shard` module. We can discuss this separately.

Fortunately, this functionality is covered in downstream tests (e.g. when we check that the sharded parameters are constructed as expected).